asab/web/authz/service.py

```diff
@@ -72,62 +93,93 @@ async def authorize(self, resources, access_token, tenant=None):
-			if access_token is not None:
-				headers["Authorization"] = "Bearer {}".format(access_token)
-
-			async with session.get(
-				userinfo_url,
-				headers=headers,
-			) as response:
-				if response.status != 200:
-					L.error("Failed to fetch userinfo", struct_data={
-						"status": response.status,
-						**dict(response.headers),
-					})
-					return None
-
-				response_json = await response.json()
-				if len(response_json) > 0:
-					userinfo = response_json
-
-				self._set_to_cache(access_token, userinfo)
-
-		return userinfo
-
-	async def _evaluate_expiration_in_cache(self, event_name):
-		cache_keys_to_delete = list()
-
-		for cache_key, cache_value in self.Cache.items():
-			authorized, expiration = cache_value
-			if self.App.time() >= expiration:
-				cache_keys_to_delete.append(cache_key)
-
-		for cache_key in cache_keys_to_delete:
-			del self.Cache[cache_key]
-
-	def _get_from_cache(self, access_token):
-		key = access_token.encode("utf-8")
-		hashed = hashlib.sha256(key).digest()
-		cache_value = self.Cache.get(hashed)
-		if cache_value is None:
+	async def userinfo(self, bearer_token):
+		if not self.is_ready():
+			L.error("AuthzService is not ready: No public keys loaded yet.")
 			return None
 
-		return cache_value[0]
+		if self._DisableTokenVerification:
+			return self._get_id_token_claims_without_verification(bearer_token)
+		else:
+			return self._get_id_token_claims(bearer_token)
 
-	def _set_to_cache(self, access_token, userinfo):
-		key = access_token.encode("utf-8")
-		hashed = hashlib.sha256(key).digest()
-		self.Cache[hashed] = (userinfo, self.App.time() + self.CacheExpiration)
+
+	def _get_id_token_claims(self, bearer_token: str):
+		assert jwcrypto is not None
+		try:
+			token = jwcrypto.jwt.JWT(jwt=bearer_token, key=self.PublicKey)
+		except jwcrypto.jwt.JWTExpired:
+			raise asab.exceptions.NotAuthenticatedError("ID token expired.")
+		except jwcrypto.jws.InvalidJWSSignature:
+			raise asab.exceptions.NotAuthenticatedError("Invalid ID token signature.")
+		except ValueError as e:
+			raise asab.exceptions.NotAuthenticatedError("Authentication failed: {}".format(e))
+
+		try:
+			token_claims = json.loads(token.claims)
+		except ValueError:
+			raise asab.exceptions.NotAuthenticatedError("Cannot parse ID token claims.")
+
+		return token_claims
+
+
+	def _get_id_token_claims_without_verification(self, bearer_token: str):
+		try:
+			header, payload, signature = bearer_token.split(".")
+		except IndexError:
+			raise asab.exceptions.NotAuthenticatedError("Cannot parse ID token: Wrong number of '.'.")
+
+		try:
+			claims = json.loads(base64.b64decode(payload.encode("utf-8")))
+		except binascii.Error:
+			raise asab.exceptions.NotAuthenticatedError("Cannot parse ID token: Payload is not base 64.")
+		except json.JSONDecodeError:
+			raise asab.exceptions.NotAuthenticatedError("Cannot parse ID token: Payload cannot be parsed as JSON.")
+
+		return claims
+
+
+	async def _get_public_keys(self):
```

Comment:
`_fetch_public_keys()` ... it is not a getter, it downloads keys.

Reply:
renamed